### ISSUE TITLE
Completions in VS Code

### ISSUE THREAD

**Issue by zachallaun** [lexical-lsp/lexical#400]
While doing some work on completions, I found a number of VS Code issues that affect completions pretty significantly. The workarounds for these may need to occur both here and in [lexical-lsp/vscode-lexical](https://github.com/lexical-lsp/vscode-lexical), so I'm creating a single tracking issue to consolidate information for now.

---

When a completion request is issued by the client, the response from the server can optionally set an `isIncomplete` flag. If it's set to `true`, according to the spec[^1]: "This list is not complete. Further typing should result in recomputing this list." This means that further typing could add/reorder items.

Without `isIncomplete` set to `true`, VS Code makes the assumption that the initial set of items is complete, and further characters in the same "word" are filtered client-side. This means that for each word, only a single completion request will be issued. This led to some completions never appearing in VS Code, fixed by #398.

Unfortunately, this fix isn't complete: VS Code has what I think must be considered a bug[^2]: If a completion result is empty, i.e. the list of completion items is `[]`, *even if `isIncomplete` is set to `true`*, VS Code will not issue additional completion requests. So if we return no completions on the first character, VS Code will *never* show additional completions from the server.

There have been additional issues and suggestions[^3][^4] for how VS Code's behavior should be improved, but I think all of them hinge on the original bug being resolved.

---

Looking through the linked issues, multiple language servers have tried various workarounds. For now, however, it may be necessary to move away from using `isIncomplete` if possible and always return the full set of completions on the first character.

[^1]: [LSP Specification: `textDocument/completion`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_completion)
[^2]: microsoft/vscode#155738
[^3]: microsoft/vscode#168329
[^4]: microsoft/vscode#147830

**Comment by scottming** [comment]
I found that the fix in PR 398 has caused some new issues. In my understanding, `is_incomplete: true` means that the server did not return all the data, and the client needs to continue making requests. This leads to more frequent communication between them and complicates the situation.

For example, in Neovim, there is now frequent flickering when completing structs, which was not the case before.

 ### now
 
https://github.com/lexical-lsp/lexical/assets/12830256/c96155a6-bdb9-4971-a627-5c2d1022b61c

### Before(Better UX)

https://github.com/lexical-lsp/lexical/assets/12830256/86221e32-5d4a-48d5-ba47-4717a4f7fb5c


Regarding the thing you mentioned in reference [^2], I also think it is a bug because other clients can handle this aspect well. Therefore, I also think we should not return all completion results in the same format with `    Completion.List.new(items: items, is_incomplete: true)
`.

**Comment by zachallaun** [comment]
You're correct, @scottming -- this is a UX regression from #398. If we start returning completions for single-character prefixes (#410 does this for all completions except atoms), we can revert #398.

The problem is that, for VS Code, it seems to be all or nothing: return all possible completions on the first character, or no completions will be shown at all for subsequent input until a trigger character or new word is seen.

**Comment by zachallaun** [comment]
Behavior in VS Code was much improved by #410, but I'd like to leave this issue open to make it hopefully easier to track the upstream issues in VS Code and as context for people coming here to report other potential issues with VS Code completions.

**Comment by scohen** [comment]
@zachallaun I don't like long-lived issues, unless they're tracking several sub-issues. Can we either close this or provide a set of tasks that when completed would close this issue?

**Comment by scohen** [comment]
Thank you!
